### PR TITLE
Damian/socketio

### LINKED ISSUE #54: Use WebSockets
Use WebSockets (Using Socket.io package) to establish 2 way connection between client and server, for real-time chat.

- [ ] Implement WebSockets on client
- [ ] Implement WebSockets on server

### PR DESCRIPTION
**Pull request by D-Garnes**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/BoopChat/boop/pulls) the bugtracker for similar pull requests
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] Covered the code with tests that prove my fix is effective or that my feature works (note that PRs
without tests will be REJECTED)
- [ ] New and existing unit tests pass locally with my changes

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.
Please also list any relevant details for your test configuration

---

### What is the purpose of your *pull request*?
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Description of your *pull request* and other information
Client-side:
	- Added socket.io-client dependency.
	- Changed client to update conversation list using the socket.
	- Changed client to update new messages using the socket.
Server-side
	- Added socket.io dependency.
	- Change conversation controller to emit new conversations using the socket.
	- Changed messages controller to emit new messages using the socket.

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains
the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much
context and examples as possible.

closes #54

### CODE REVIEW

File: client/src/components/messenger/Conversations.js [hunk 1]
```diff
@@ -50,66 +52,76 @@ const AddConversationDialog = ({ open, onClose, token }) => {
...
                 let { title, list } = conversationDetails;
-                let result = await ConversationsController.createConversation(token, list, title);
-                if (result.success) { // add new conversation to the back of the list
-                    let { conversation } = result;
-                    setConversations(conversations.length > 0 ? [...conversations, conversation]: [conversation]);
-                    // auto open chat after creation
-                    selectConversation(conversation.id, conversation.title);
+                if (title.length < 1) {
+                    // create a title from the chosen participants' names
+                    let chop = list.length < 2 ? 20 : list.length < 3 ? 9 : 5;
```
Comment (davidfowl) [inline client/src/components/messenger/Conversations.js:113]: Maybe but these numbers in constants?

File: client/src/components/messenger/Messenger.js [hunk 2]
```diff
@@ -16,36 +17,57 @@ const Messenger = () => {
...
+        setCurrentConvo({ id, title, participants });
     };
 
     const userInfo = useSelector((state) => state.user.userInfo);
+    const token = useSelector((state) => state.user.token);
 
     const updateUser = (userInfo) => dispatch(setUserInfo(userInfo));
 
+    const [socket] = useState(() => {
+        let soc = io("http://localhost:5000", {
```
Comment (davidfowl) [inline client/src/components/messenger/Messenger.js:30]: Why do we need the full address here? This should be relative no?

File: client/src/components/messenger/Messenger.js [hunk 3]
```diff
@@ -16,36 +17,57 @@ const Messenger = () => {
...
+            auth: {
+                token,
+            },
+        });
+        soc.on("connect", () => console.log("connect: ", soc.id));
+        soc.on("test", (message) => console.log(message));
```
Comment (davidfowl) [inline client/src/components/messenger/Messenger.js:36]: Remove?

File: client/src/components/messenger/Messenger.js [hunk 4]
```diff
@@ -16,36 +17,57 @@ const Messenger = () => {
...
+        soc.on("connect_error", (error) => console.log(error));
```
Comment (davidfowl) [inline client/src/components/messenger/Messenger.js:37]: console.error?

File: client/src/components/messenger/Conversations.js (hunk 1 above)
Comment (JChris246) [reply client/src/components/messenger/Conversations.js]: This was probably a bad idea in the first place. For example, if the user didn't provide a title for the chat and the participant list was ["Melissa", "Krystal", "Christopher"], the title would end up being: "You, Melis, Kryst, Chris", which is not the best looking or could be confusing

File: client/src/components/messenger/Messenger.js (hunk 2 above)
Comment (JChris246) [reply client/src/components/messenger/Messenger.js:30]: @D-Garnes

Comment (D-Garnes) [reply client/src/components/messenger/Messenger.js:30]: The current react proxy is not working with the socket.io that is why the full address was used. I'll sort it out.

**Review by JChris246 — APPROVED** [review]
👍

**Pull request merged**